tfprotov5/tftypes/walk.go

```diff
@@ -0,0 +1,209 @@
+package tftypes
+
+// these functions are based heavily on github.com/zclconf/go-cty
+// used under the MIT License
+//
+// Copyright (c) 2017-2018 Martin Atkins
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import (
+	"errors"
+)
+
+// Walk traverses a Value, calling the passed function for every element and
+// attribute in the Value. The AttributePath passed to the callback function
+// will identify which attribute or element is currently being surfaced by the
+// Walk, and the passed Value will be the element or attribute at that
+// AttributePath. Returning true from the callback function will indicate that
+// any attributes or elements of the surfaced Value should be walked, too;
+// returning false short-circuits the walk at that element or attribute, and
+// does not visit any of its descendants. The return value of the callback does
+// not matter when the Value that has been surfaced has no elements or
+// attributes. Walk uses a depth-first traversal.
+func Walk(val Value, cb func(AttributePath, Value) (bool, error)) error {
+	var path AttributePath
+	return walk(path, val, cb)
+}
+
+func walk(path AttributePath, val Value, cb func(AttributePath, Value) (bool, error)) error {
+	shouldContinue, err := cb(path, val)
+	if err != nil {
+		var ape attributePathError
+		if !errors.As(err, &ape) {
+			err = path.NewError(err)
+		}
+		return err
+	}
+	if !shouldContinue {
```

Comment:
Thought experiment: I wonder if this should somehow surface that the visitation was interrupted before a complete exploration of all nodes could be finished. A special type of error maybe?

Reply:
This seems to mirror the Go standard library walk behaviors, in the expectation that the original caller has some control over the callback? And given the callback is the one returning `shouldContinue`, they should know some things may be skipped? Similarly, the callback can return its own error that the caller can handle. I guess I'm not clear on when this would be used, or why.